packages/devtools_app/lib/src/timeline/flutter/flutter_frames_chart.dart

```diff
@@ -2,20 +2,669 @@
+      if (timerDataIndex < _cannedData.length) {
+        _frameDurations.add(createBarEntry(timerDataIndex));
+        timerDataIndex += 3;
+      }
+
+      updateChart();
+    });
+  }
+
+  void updateChart() {
+    _chartController.data = BarData([]..add(frameDurationsSet));
+
+    setState(() {
+      // Signal data has changed.
+      frameDurationsSet.notifyDataSetChanged();
+    });
+  }
+
+  void _loadAllData(int startIndex) {
+    int index = startIndex;
+    while (index < _cannedData.length) {
+      _frameDurations.add(createBarEntry(index));
+      index += 3;
+    }
+  }
+
   @override
   Widget build(BuildContext context) {
-    return Padding(
-      padding: const EdgeInsets.only(bottom: 8.0),
-      child: Container(
-        height: 150.0,
-        color: Colors.black26,
-        child: const Center(
-          child: Text('TODO Flutter Frames Chart'),
-        ),
+    return Container(
+      height: 200.0,
+      child: BarChart(_chartController),
+    );
+  }
+}
+
+class YAxisMilliSecFormatter extends ValueFormatter {
+  @override
+  String getFormattedValue1(double value) {
+    return value.toInt().toString();
+  }
+}
+
+typedef SelectionCallback = void Function(int frameIndex);
+
+/*
+ * Selection of a point in the Bar chart displays the data point values
+ * UI duration and GPU duration. Also, highlight the selected stacked bar.
+ * 
+ * onSelected callback function invoked when bar entry is selected.
+ */
+class CustomDataValues extends LineChartMarker {
+  CustomDataValues({
+    this.textColor,
+    this.backColor,
+    this.fontSize,
+    this.onSelected,
+  }) {
+    _formatter = DefaultValueFormatter(2);
+    textColor ??= ColorUtils.WHITE;
+    backColor ??= const Color.fromARGB(127, 0, 0, 0);
+    fontSize ??= 10;
+//    fontSize ??= Utils.convertDpToPixel(2);
+  }
+
+  Entry _entry;
+  Highlight _highlight;
+
+  DefaultValueFormatter _formatter;
+  Color textColor;
+  Color backColor;
+  double fontSize;
+
+  int _lastFrameIndex = -1;
+  final SelectionCallback onSelected;
+
+  @override
+  void draw(Canvas canvas, double posX, double posY) {
+    final frameIndex = _entry.x.toInt();
+    final yValues = (_entry as BarEntry).yVals;
+
+    final num uiDuration = yValues[1];
+    final num gpuDuration = yValues[0];
+
+    final TextPainter painter = PainterUtils.create(
+      null,
+      'UI  = ${_formatter.getFormattedValue1(uiDuration)}\n'
+      'GPU = ${_formatter.getFormattedValue1(gpuDuration)}',
+      textColor,
+      fontSize,
+    )..textAlign = TextAlign.left;
+
+    final Paint paint = Paint()
+      ..color = backColor
+      ..strokeWidth = 2
+      ..isAntiAlias = true
+      ..style = PaintingStyle.fill;
+
+    final MPPointF offset = getOffsetForDrawingAtPoint(
+      posX,
+      posY,
+    );
+
+    canvas.save();
+    // translate to the correct position and draw
+    painter.layout();
+    final Offset pos = calculatePos(
+      posX + offset.x,
+      posY + offset.y - 15,
```

Comment:
nit: 15 seems a little magic. Put in a const and give a descriptive name describing what it is.

Reply:
Done.